src/emulator/auth/handlers.ts

```diff
@@ -130,28 +130,203 @@ export function registerHandlers(
+<title>Auth Emulator IDP Login Widget</title>
+<h1>Sign-in with <span class="js-provider-id">Provider</span></h1>
+<p>Please select an existing account in the Auth Emulator or add a new one.</p>
+<p><b>This page is a work in progress and is subject to change.</b></p>
+<div>
+  <ul>
+    <!-- TODO: Generate the list from all existing accounts with providerId. -->
+    <li class="js-reuse-account" data-id-token='{"sub":"12345"}'>
+      <a href="#">Test Account 1</a>
+    </li>
+    <li class="js-reuse-account" data-id-token='{"sub":"67890"}'>
+      <a href="#">Test Account 2</a>
+    </li>
+    <li class="js-new-account"><a href="#">Add Another Account</a></li>
+  </ul>
+</div>
+<script>
+  // TODO: Support older browsers where URLSearchParams is not available.
+  var query = new URLSearchParams(location.search);
+  var apiKey = query.get('apiKey');
+  var appName = query.get('appName');
+  var authType = query.get('authType');
+  var providerId = query.get('providerId');
+
+  var redirectUrl = query.get('redirectUrl');
+  var scopes = query.get('scopes');
+  var eventId = query.get('eventId');
+  if (!apiKey || !appName || !authType || !providerId) {
+    alert('Auth Emulator Internal Error: Missing query params apiKey / appName / authType / providerId.');
+  }
+  var storageKey = apiKey + ':' + appName;
+
+  function saveAuthEvent(authEvent) {
+    if (/popup/i.test(authType)) {
+      sendAuthEventViaIframeRelay(authEvent, function (err) {
+        if (err) {
+          return alert('Auth Emulator Internal Error: ' + err);
+        }
+      });
+    } else {
+      saveAuthEventToStorage(authEvent);
+      if (redirectUrl) {
+        window.location = redirectUrl;
+      } else {
+        // TODO
+        return alert('This feature is not implemented in Auth Emulator yet. Please use signInWithCredential for now.');
+      }
+    }
+  }
+
+  function saveAuthEventToStorage(authEvent) {
+    sessionStorage['firebase:redirectEvent:' + storageKey] =
+      JSON.stringify(authEvent);
+  }
+
+  function sendAuthEventViaIframeRelay(authEvent, cb) {
+    var sent = false;
+    if (window.opener) {
+      for (var i = 0; i < window.opener.frames.length; i++) {
+        var iframeWin = window.opener.frames[i];
+        var query = new URLSearchParams(iframeWin.location.search);
+        if (query.get('apiKey') === apiKey && query.get('appName') === appName) {
+          iframeWin.postMessage({
+            data: {authEvent: authEvent, storageKey: storageKey},
+            eventId: Math.floor(Math.random() * Math.pow(10, 20)).toString(),
+            eventType: "sendAuthEvent",
+          }, '*');
+          sent = true;
+        }
+      }
+    }
+    if (!sent) {
+      return cb('No matching frame');
+    }
+    return cb();
+  }
+
+  // DOM logic
+
+  document.querySelector('.js-provider-id').textContent = providerId;
+  var reuseAccountEls = document.querySelectorAll('.js-reuse-account');
+  [].forEach.call(reuseAccountEls, function (el) {
```

Comment:
Isn't this a noop?

Reply:
This is a crazy hack from the old days of JavaScript, and what it really does is call `Array.prototype.forEach` but with `this` set to the NodeList. Note that `NodeList` is not an Array and doesn't have `.forEach`, thus the hack. See https://developer.mozilla.org/en-US/docs/Web/API/NodeList#Example

This hack was obsoleted by `for...of`. Modern browsers also added `.forEach` (but not universal, notably not in IE).